Implement margin-trim for floats in block containers that contain only block boxes.

sammygill opened the pull request:
#### f679766e27a17ddcbdd8a562f6f1581fab68d053
<pre>
Implement margin-trim for floats in block containers that contain only block boxes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249207">https://bugs.webkit.org/show_bug.cgi?id=249207</a>
rdar://103285803

Reviewed by Alan Baradlay.

Another patch will be added to provide support for block containers that
contain only inline level boxes.

There are 3 different pieces that we need to implement:

1.) Trimming the block start, inline start, and inline end margins
2.) Trimming the block-end margin, which is slightly different from
the others
3.) Making sure the trimmed margins do not contribute to the inrinsic
sizing of the container

When a candidate position for a float is determined in the float
positioning code, we can determine whether it is possible to trim its
margins. For each candidate position, we can determine whether any edges
of the margin box would touch the container and then trim that edge if
it is specified in margin-trim. This can be done for the inline-start,
inline-end, and block-start edges of the container.

Trimming the block-end margin is slightly different because it is not
in the context of the containing block like the other margins but of the
block formatting context. This means that we can only trim these margins
after its block formatting context is done with layout. In
RnderBlockFlow we can check to see if it establishes a block formatting
context and then proceed to trim the block-end margins of its floats
where needed. As we iterate over each float, we check to see if its block
end location is lower than the lowest piece of content in the BFC. If
so, we trim it up to the necesssary amount where it will not extend the
height of the BFC.

Finally, to make sure that the trimmed inline margins do not contribute
to the intrinsic sizing of its containing block, all we need to do
is check to see if the item is a float and if any of the inline margins
are specified to be trimmed. If so, we do not include those in its
contribution.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-deeply-nested-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-deeply-nested.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers-vert-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-same-container-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-same-container.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-relative-positioned-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-relative-positioned.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-with-transforms-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-with-transforms.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-fit-content-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-fit-content-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout-vert-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-with-block-content-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-with-block-content-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-right-trimmed-margin-allows-float-to-fit-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-right-trimmed-margin-allows-float-to-fit-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-overflowing-float-margins-tirmmed-at-final-position-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-overflowing-float-margins-tirmmed-at-final-position-block-layout.html: Added.
* Source/WebCore/rendering/FloatingObjects.cpp:
(WebCore::FloatingObject::isLowestPlacedFloatBottomInBlockFormattingContext const):
* Source/WebCore/rendering/FloatingObjects.h:
(WebCore::FloatingObject::setHeight):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeBlockPreferredLogicalWidths const):
* Source/WebCore/rendering/RenderBlock.h:
(WebCore::RenderBlock::logicalMarginBoxHeightForChild const):
(WebCore::RenderBlock::logicalMarginBoxTopForChild const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock):
(WebCore::RenderBlockFlow::trimFloatBlockEndMargins):
(WebCore::RenderBlockFlow::shouldChildInlineMarginContributeToContainerIntrinsicSize const):
(WebCore::RenderBlockFlow::insertFloatingObject):
(WebCore::RenderBlockFlow::trimMarginForFloat):
(WebCore::RenderBlockFlow::computeLogicalLocationForFloat):
(WebCore::RenderBlockFlow::blockFormattingContextInFlowBlockLevelContentHeight const):
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::establishesBlockFormattingContext const):
(WebCore::RenderBox::blockFormattingContextRoot const):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: <a href="https://commits.webkit.org/260318@main">https://commits.webkit.org/260318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fdb75fc05ba8a6093913cec82d8b492d7caf6d4e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117018 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116368 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111769 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18330 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8254 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100063 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113643 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13834 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97048 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41541 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28686 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9866 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30034 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6929 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49622 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7138 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12144 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4970,6 +4970,24 @@ bool RenderBox::establishesIndependentFormattingContext() const
+    if (!boxStyle.isDisplayBlockLevel())
+        return false;
+    auto isBlockWithOverFlowOtherThanVisibleAndClip = [&]() {
+        auto boxOverflowX = boxStyle.overflowX();
```
should this use effectiveOverflow* instead?

sammygill replied:
As far as I can tell there is no reason that it should not. Went ahead and changed it to use it

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5630,4 +5648,14 @@ std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerHeight() const
+    const auto container = containingBlock();
+    if (!container)
+        return nullptr;
+    return container->createsBlockFormattingContext() ? dynamicDowncast<RenderBlockFlow>(container) : container->blockFormattingContextRoot(); 
```
I try to avoid recursive calls when possible (this is essentially a loop on the ancestor chain until you hit the first BFC root)

sammygill replied:
Changed to iterate over the elements instead of recursing

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.h:
```diff
@@ -24,6 +24,7 @@
 
 #include "CaretRectComputation.h"
 #include "FloatingObjects.h"
+#include "HTMLHtmlElement.h"
```
?

sammygill replied:
Removed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.h:
```diff
@@ -226,6 +227,8 @@ class RenderBlockFlow : public RenderBlock {
 
     bool shouldTrimChildMargin(MarginTrimType, const RenderBox&) const final;
+    void trimFloatBlockEndMargins(LayoutUnit blockFormattingContextInFlowContentHeight);
+    bool shouldChildInlineMarginContributeToContainerIntrinsicSize(MarginTrimType, const RenderObject&) const;
```
I am sure you can't call this with RenderText, so RenderElement would be a better type.

sammygill replied:
Changed!

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4663,5 +4743,18 @@ bool RenderBlockFlow::tryComputePreferredWidthsUsingModernPath(LayoutUnit& minLo
+{
+    ASSERT(createsBlockFormattingContext());
+    if (childrenInline())
+        return 0_lu;
```
BFC root can't have inline children?

sammygill replied:
Removed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4663,5 +4743,18 @@ bool RenderBlockFlow::tryComputePreferredWidthsUsingModernPath(LayoutUnit& minLo
+    if (!lastChild)
+        return 0_lu;
+    // Child's margin box starting location + border box height + margin after size
+    return lastChild->y() + lastChild->height() + lastChild->marginAfter(&style());
```
don't we have computed margin box geometry by the time we get here?

alanbaradlay replied:
how does this work with vertical writing-mode?

sammygill replied:
Changed to check `isHorizontalWritingMode` first and then using `lastChild->marginBoxRect().height/width`

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1375,50 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+            startLocation += containingBlock->y();
+            containingBlock = containingBlock->containingBlock();
+        }
+        return startLocation + floatBox.height();
```
How does this work with transforms and other non-trivial translates like position: relative? is this writing-mode aware?

alanbaradlay replied:
(though maybe the fact that we stay within BFC makes it a lot simpler than something like in localToContainerQuad/mapLocalToContainer, I'd still check out those functions as you do a very similar translate here.)

sammygill replied:
Changed to use `logicalTopForChild` and `logicalHeightForChild` so that we get correct value according to the containing block's writing mode

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -69,6 +69,7 @@
 #include "TextAutoSizing.h"
 #include "VerticalPositionCache.h"
 #include "VisiblePosition.h"
+#include "html/HTMLHtmlElement.h"
```
?

sammygill replied:
Removed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4663,5 +4755,17 @@ bool RenderBlockFlow::tryComputePreferredWidthsUsingModernPath(LayoutUnit& minLo
+{
+    ASSERT(createsBlockFormattingContext());
+    // For block layout we should just be able to check the height of the last in flow box
+    auto lastChild = lastInFlowChildBox();
```
how does this work for simple cases like `<div bfcroot>this does not have a render box only a render text</div>`?

sammygill replied:
Oh I think I misunderstood your previous comment. Yes I agree we should be able to get the content height from the inline content. I actually have this logic in my other patch (https://github.com/WebKit/WebKit/pull/9145/files#diff-8ddefc45aa5ce620250d06c1512ae9a263aaa9af625660ca8d50312e750c19f2) but I can bring it into this one as well.

Since I wanted to split it up into two patches (one to trim the floats in an IFC), I didn't add it here. Perhaps instead I can add a check `!hasChildrenInline` before calling `trimFloatBlockEndMargins` inside `RenderBlockFlow::layoutBlock` and then remove it in the IFC patch?

alanbaradlay replied:
I am sorry if my previous comment was misleading.
I think when you add a function called "foobar" it should do foobar, exactly what the name implies. At this point this function returns the height of the block level content and it should be named accordingly. When your upcoming patch changes the behavior, you should rename it. It helps to make sure nothing gets out-of-sync when reverts happen (i.e we don't end up with a function called blockFormattingContextInFlowContentHeight() with block level elements only and some other code starts calling it assuming it returns the height of the inflow content).

sammygill replied:
Yeah that makes sense. I'll be more aware of making sure I changed function names and not only the logic associated with them if they get changed like this in future patches.

sammygill replied:
Changed to `blockFormattingContextInFlowBlockLevelContentHeight`

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4663,5 +4755,17 @@ bool RenderBlockFlow::tryComputePreferredWidthsUsingModernPath(LayoutUnit& minLo
+        return 0_lu;
+    auto lastChildMarginBoxLogicalHeight = isHorizontalWritingMode() ? lastChild->marginBoxRect().height() : lastChild->marginBoxRect().width();
+    // Child's margin box starting location + border box height + margin after size
+    return lastChild->marginBoxRect().y() + lastChildMarginBoxLogicalHeight;
```
is marginBoxRect().y() what you expect to be the correct value in vertical mode?

sammygill replied:
Nope. I created `logicalMarginBoxTopForChild`and `logicalMarginBoxHeightForChild` to help with this so should return the correct values according to containing block's writing mode

alanbaradlay commented on Source/WebCore/rendering/FloatingObjects.h:
```diff
@@ -64,7 +64,7 @@ class FloatingObject {
     void setY(LayoutUnit y) { ASSERT(!isInPlacedTree()); m_frameRect.setY(y); }
     void setWidth(LayoutUnit width) { ASSERT(!isInPlacedTree()); m_frameRect.setWidth(width); }
-    void setHeight(LayoutUnit height) { ASSERT(!isInPlacedTree()); m_frameRect.setHeight(height); }
+    void setHeight(LayoutUnit height) { ASSERT(!isInPlacedTree() || containingBlockHasBlockEndMarginTrim()); m_frameRect.setHeight(height); }
```
I think this needs a bit stronger assertion. What we need to assert here is that changing the height of this float box has no impact on the rest of the floats.

sammygill replied:
I changed it instead to `isLowestPlacedFloatBottomInContainingBlock` since that is the only time we should be trimming those margins.

alanbaradlay commented on Source/WebCore/rendering/FloatingObjects.h:
```diff
@@ -92,6 +92,10 @@ class FloatingObject {
     void setOriginatingLine(LegacyRootInlineBox& line) { m_originatingLine = line; }
 
+    bool containingBlockHasBlockEndMarginTrim() const;
+    void setHasTrimmedMargins(bool hasTrimmedMargins) { m_hasTrimmedMargins = hasTrimmedMargins; }
```
This could be just a simple setHasTrimmedMargins() { m_hasTrimmedMargins = true; } since you don't go from trimmed to untrimmed.

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/FloatingObjects.h:
```diff
@@ -110,6 +114,7 @@ class FloatingObject {
     unsigned m_paintsFloat : 1;
     unsigned m_isDescendant : 1;
     unsigned m_isPlaced : 1;
+    unsigned m_hasTrimmedMargins : 1;
```
Where do you initialize this member?

sammygill replied:
Thanks for catching this. Initialized in the constructors now

alanbaradlay replied:
FloatingObject seems to have multiple constructors and I am only seeing one initialization here.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5630,4 +5648,14 @@ std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerHeight() const
+    while (cb && !cb->createsBlockFormattingContext())
+        cb = cb->containingBlock();
+    return dynamicDowncast<RenderBlockFlow>(cb);
+}
```
I can't find any caller of this function.

sammygill replied:
Should be used now with some other changes

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -705,6 +706,7 @@ class RenderBox : public RenderBoxModelObject {
     }
 
     bool establishesIndependentFormattingContext() const override;
+    bool createsBlockFormattingContext() const;
```
Any reason why you chose "creates" instead of the spec term of "establishes" (same as the function above)?

sammygill replied:
Nope. Changed.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -192,10 +192,30 @@ RenderBlockFlow* RenderBlockFlow::previousSiblingWithOverhangingFloats(bool& par
+    // 2.) Those margins will get trimmed at some point during layout (which gets reflected by seting the m_marginBox values to 0)
+    // 3.) A second iteration of layout will occur and RenderBlockFlow will clear its FloatingObjects
+    // 4.) The floating objects will be created with the wrong width since they grab the trimmed margins from the m_marginBox
+    auto& floatingObjectSet = m_floatingObjects->set();
```
what does first and second iteration of layout mean here? Block layout is a one pass type of layout (unless you mean paginated content here -btw, I am seeing any test with pagination. Does trimming get readjusted when balancing columns?)

sammygill replied:
The issue was occurring when I was noticing that the page went through the initial layout via `FrameViewLayoutContext::performLayout`. After this was completed the call to `FrameView::adjustViewSize` recursed back into `FrameViewLayoutContext::performLayout.` The floats did not seemed to be marked for layout so the calls to `layoutIfNeeded` did nothing. Is there a better phrase I should use here?

alanbaradlay replied:
ok, so this second layout initiated by adjustViewSize should not need any special treatment in theory. If constraints changed between the initial and the second layout, renderers should be marked dirty accordingly, If not the floats here should not be laid out again.

sammygill replied:
Right I agree with that so I'm not entirely sure what was happening in the scenarios. The containing block was going through layout again just normally, but when it recreated the `FloatingObjects` it looked like those renderers were not being dirtied which seems incredibly strange? Maybe this is something we could create a new bug for?

alanbaradlay replied:
yes, pleas file a bug, mark the affected test(s) fail and remove this. We probably need a non-trimming specific fix on this.

sammygill replied:
New bug: https://bugs.webkit.org/show_bug.cgi?id=252183

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1394,51 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+    };
+    for (auto& floatingObject : m_floatingObjects->set()) {
+        auto& floatBox = floatingObject->renderer();
+        if (!floatBox.parentStyle()->marginTrim().contains(MarginTrimType::BlockEnd) || !floatBox.marginAfter())
```
is it parent or containing block?

sammygill replied:
Containing block. Thanks for catching

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1394,51 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+        while (!containingBlock->createsBlockFormattingContext()) {
+            startLocation += logicalTopForChild(*containingBlock);
+            containingBlock = containingBlock->containingBlock();
+        }
```
don't we have a helper on this (translate some coords relative to an ancestor)?

alanbaradlay replied:
Also please add deeply nested type of test cases, where the float is a descendent of block boxes with both layout and paint type of offsets (e.g. border, padding and relatively positioned, transformed etc).

sammygill replied:
Added a few of those tests. Also after talking with Simon it seems like using `localToAbsolute` is reasonable to use while walking up the containing block chain here

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4387,8 +4521,8 @@ void RenderBlockFlow::computeInlinePreferredLogicalWidths(LayoutUnit& minLogical
-                    Length endMargin = childStyle.marginEndUsing(&style());
+                    Length startMargin = shouldChildInlineMarginContributeToContainerIntrinsicSize(MarginTrimType::InlineStart, *dynamicDowncast<RenderElement>(child)) ? childStyle.marginStartUsing(&style()) : Length(0, LengthType::Fixed); 
+                    Length endMargin = shouldChildInlineMarginContributeToContainerIntrinsicSize(MarginTrimType::InlineEnd, *dynamicDowncast<RenderElement>(child)) ? childStyle.marginEndUsing(&style()) : Length(0, LengthType::Fixed);
                     if (startMargin.isFixed())
```
I can't figure out what's going on up here ^^^ so I am just going to trust that we've got enough coverage on these lines.

alanbaradlay commented on Source/WebCore/rendering/FloatingObjects.cpp:
```diff
@@ -112,6 +113,13 @@ LayoutSize FloatingObject::translationOffsetToAncestor() const
     return locationOffsetOfBorderBox() - renderer().locationOffset();
 }
 
+bool FloatingObject::isLowestPlacedFloatBottomInContainingBlock() const
```
you may wanna put this behind #ifdef if only used in ASSERT

sammygill replied:
Done

alanbaradlay commented on Source/WebCore/rendering/FloatingObjects.h:
```diff
@@ -64,7 +64,7 @@ class FloatingObject {
     void setY(LayoutUnit y) { ASSERT(!isInPlacedTree()); m_frameRect.setY(y); }
     void setWidth(LayoutUnit width) { ASSERT(!isInPlacedTree()); m_frameRect.setWidth(width); }
-    void setHeight(LayoutUnit height) { ASSERT(!isInPlacedTree()); m_frameRect.setHeight(height); }
+    void setHeight(LayoutUnit height) { ASSERT(!isInPlacedTree() || isLowestPlacedFloatBottomInContainingBlock()); m_frameRect.setHeight(height); }
```
I am not sure if "inContainingBlock" adds any value and it is rather misleading since you can have the lowest placed float in a containing block and yet not the lowest in the BFC which is what your trimming context is (actually this needs to be fixed in the function above) .

sammygill replied:
I changed it to "inBlockFormattingContext" to distinguish it from absolute coordinates, but let me know if you want me to change it. I also changed the implementation of the method to use the BFC root as well.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+        while (!containingBlock->establishesBlockFormattingContext()) {
+            startLocation += containingBlock->localToAbsolute();
+            containingBlock = containingBlock->containingBlock();
+        }
```
I am bit puzzled about accumulating absolute coordinates here. To my understanding localToAbsolute() returns the box's absolute coordinate, relative to RenderView.

sammygill replied:
I stepped through the calls and it seems like `RenderBox::mapLocalToContainer` gets nullptr for `ancestorContainer` and `view().frameView().layoutContext().isPaintOffsetCacheEnabled()` is true. In that case it looks like it uses `locationOffset` and returns early. Later parts of the code seem to call up the containers, though. 

I'm not super familiar with the intricacies of this function, so definitely might not be the right thing to use. Maybe instead we could use `offsetFromContainer` until we get up to the BFC root instead? Thoughts?

sammygill replied:
Changed to use `offsetFromContainer` instead

alanbaradlay commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2349,8 +2349,17 @@ void RenderBlock::computeBlockPreferredLogicalWidths(LayoutUnit& minLogicalWidth
-        Length endMarginLength = childStyle.marginEndUsing(&styleToUse);
+        Length startMarginLength;
+        Length endMarginLength;
+        if (auto blockBox = dynamicDowncast<RenderBlockFlow>(this)) {
```
why is this limited to RenderBlockFlow -any particular reason why shouldChildInlineMarginContributeToContainerIntrinsicSize is on RenderBlockFlow?

sammygill replied:
I added it to that because of this part of the spec: 

`It also affects floats for which the block container is a containing block by: Zeroing the inline-axis margins of a float for the purpose of calculating its intrinsic size contributions or its size in the container’s inline axis.`

But maybe the interface here could be a little better. One thing we could do is maybe make this a virtual method inside `RenderBox` and have the `RenderBlockFlow` version override it?

alanbaradlay replied:
It's fine, it's just odd that you call a base class function and immediately downcast to the derived class you called from (and yes, it should be a virtual function on the RenderBlock and RenderBlockFlow would override it).

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+    };
+    for (auto& floatingObject : m_floatingObjects->set()) {
+        auto& floatBox = floatingObject->renderer();
+        if (!floatBox.containingBlock()->style().marginTrim().contains(MarginTrimType::BlockEnd) || !floatBox.marginAfter())
```
what happens to negative margins?

sammygill replied:
I was having a hard time finding a test case where negative margins for a float increased the height of the BFC. Is that possible? For now I changed it to skip the float if the margin is <= 0

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+        if (floatMarginAfterStartLocation >= blockFormattingContextInFlowContentHeight)
+            trimMarginForFloat(*floatingObject, MarginTrimType::BlockEnd, floatBox.marginAfter(&style()));
+        else if (blockFormattingContextInFlowContentHeight > floatMarginAfterStartLocation && (blockFormattingContextInFlowContentHeight < floatMarginAfterStartLocation + floatBox.marginAfter()))
+            trimMarginForFloat(*floatingObject, MarginTrimType::BlockEnd, floatMarginAfterStartLocation + floatBox.marginAfter(&style()) - blockFormattingContextInFlowContentHeight);
```
I don't think you need to pass in the box's own style to marginAfter.

sammygill replied:
Just to make sure, does box here refer to the containing block? That's what I was trying to do so that the correct margin was retrieved using the writing mode of the containing block

alanbaradlay replied:
ok, it was super not obvious (this is a terrible interface). Also, you don't seem to be calling it with the containing block's style but with the BFC root's style (which in practice when it comes to figuring out the writing-mode does not make any difference, but as you can see I made a mistake as the reviewer and I think you did as the author, so we should probably refrain from using it and instead do something more explicit).

sammygill replied:
Oh wow good catch. I went ahead and stored the float's containing block into its own variable since we use it multiple times here. With that I can use the `marginAfterForChild` on `RenderBlock` which should get the appropriate value. It's a little more explicit since we are using the containing block directly.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -2517,6 +2593,32 @@ LayoutUnit RenderBlockFlow::logicalRightOffsetForPositioningFloat(LayoutUnit log
+    switch (marginTrimType) {
+    case MarginTrimType::InlineStart: {
+        setLogicalWidthForFloat(floatingObject, logicalWidthForFloat(floatingObject) - floatingObject.renderer().marginStart(&style()));
+        floatingObject.renderer().setMarginStart(0_lu, &style());
```
FloatingObject has a setMarginOffset() function. Do you know if these need to be in sync?

sammygill replied:
Yes I think so. The margin offset for the float are being set after we trim them, `RenderBlockFlow::computeLogicalLocationForFloat` so we should be good for that case I believe. I also added an assert `ASSERT(!floatingObject.isPlaced() || marginTrimType == MarginTrimType::BlockEnd)` inside `RenderBlockFlow::trimMarginForFloat` so that this method does not accidentally get called after a float gets its position and gets out of sync with the margin offset

alanbaradlay replied:
👍

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5630,4 +5648,14 @@ std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerHeight() const
+{
+    // This method should probably not be called on the initial containing block
+    ASSERT(!is<HTMLHtmlElement>(element()));
+    auto cb = containingBlock();
```
please try not to abbreviate unless absolutely necessary.

sammygill replied:
Fixed

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -185,6 +185,7 @@ class RenderBox : public RenderBoxModelObject {
     FloatRect repaintRectInLocalCoordinates() const override { return borderBoxRect(); }
     FloatRect objectBoundingBox() const override { return borderBoxRect(); }
 
+    RenderBlockFlow* blockFormattingContextRoot() const;    
```
Can this RenderBlockFlow be const?

sammygill replied:
I believe so. Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+    // For the purposes of computing a block container's intrinsic size, a float should not
+    // include its trimmed margins in its intrinsic size contributions
+    ASSERT(marginSide == MarginTrimType::InlineStart || marginSide == MarginTrimType::InlineEnd);
+    return !child.isFloating() || !style().marginTrim().contains(marginSide);
```
This being a public function on RenderBlockFlow, what guarantees that the child's containing block is *this?

sammygill replied:
Added this extra check to the assert to make sure that this condition holds

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -2544,7 +2660,10 @@ void RenderBlockFlow::computeLogicalLocationForFloat(FloatingObject& floatingObj
-        while (logicalRightOffsetForPositioningFloat(logicalTopOffset, logicalRightOffset, false, &heightRemainingRight) - floatLogicalLeft < floatLogicalWidth) {
+        auto availableSpace = logicalRightOffsetForPositioningFloat(logicalTopOffset, logicalRightOffset, false, &heightRemainingRight) - floatLogicalLeft;
+        while (availableSpace < floatLogicalWidth) {
+            if (containingBlockMarginTrim.contains(MarginTrimType::InlineStart) && floatLogicalLeft == logicalLeftOffset && availableSpace >= floatWidthWithoutMargin(floatingObject, MarginTrimType::InlineStart))
```
Sometimes I find it useful to either paste the relevant parts of the spec or name variables in a way that they are easily recognizable. Something along the lines of
// It also affects floats for which the block container is a containing block by:
// Discarding the inline-start/inline-end margin of an inline-start/inline-end float 
// (or equivalent) whose outer edge on that side coincides with the inner edge of the container.
auto floatHasAdjoiningMargin = containingBlockMarginTrim.contains(MarginTrimType::InlineStart) && floatLogicalLeft == logicalLeftOffset;
if (floatHasAdjoiningMargin && availableSpace >= floatWidthWithoutMargin(floatingObject, MarginTrimType::InlineStart)

sammygill replied:
yeah that is actually really good! Added the helper and added some comments above

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -2553,13 +2672,22 @@ void RenderBlockFlow::computeLogicalLocationForFloat(FloatingObject& floatingObj
+        if (containingBlockMarginTrim.contains(MarginTrimType::InlineStart) && logicalLeftOffset == floatLogicalLeft)
+            trimMarginForFloat(floatingObject, MarginTrimType::InlineStart);
+        if (containingBlockMarginTrim.contains(MarginTrimType::InlineEnd) && floatLogicalLeft + logicalWidthForFloat(floatingObject) == logicalRightOffset)
+            trimMarginForFloat(floatingObject, MarginTrimType::InlineEnd);
```
> Discarding the inline-start/inline-end margin of an inline-start/inline-end float (or equivalent) whose outer edge on that side coincides with the inner edge of the container.

I am not super good at interpreting spec in general (so please bear with me), but I am wondering if trimming an inline-end of an inline-start float is really what we are supposed to do.

sammygill replied:
Yeah so my reasoning behind this because one of the introductory sentences for the property was: `This property allows the container to trim the margins of its children where they adjoin the container’s edges.`

However, I suppose since the authors specified the behavior this way for floats we should follow it that was

So in this case:

```
<div>
<div style="float: left; width: 50px; height: 50px; margin-inline: 50px; background-color: green;">
</div>
</div>
```
only the inline-start margin would be trimmed and the remaining inline-end margin would remain. Similar changes would need to be made for the inline layout portion as well

sammygill replied:
Ended up removing the logic so now it follows exactly what the spec says.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4387,8 +4525,8 @@ void RenderBlockFlow::computeInlinePreferredLogicalWidths(LayoutUnit& minLogical
-                    Length startMargin = childStyle.marginStartUsing(&style());
-                    Length endMargin = childStyle.marginEndUsing(&style());
+                    Length startMargin = shouldChildInlineMarginContributeToContainerIntrinsicSize(MarginTrimType::InlineStart, *dynamicDowncast<RenderElement>(child)) ? childStyle.marginStartUsing(&style()) : Length(0, LengthType::Fixed); 
+                    Length endMargin = shouldChildInlineMarginContributeToContainerIntrinsicSize(MarginTrimType::InlineEnd, *dynamicDowncast<RenderElement>(child)) ? childStyle.marginEndUsing(&style()) : Length(0, LengthType::Fixed);
```
This should be in the inline patch.

sammygill replied:
Removed

alanbaradlay reviewed (approved):
this should be good to go (after that &style() adjustment).

sammygill committed:
Implement margin-trim for floats in block containers that contain only block boxes.
https://bugs.webkit.org/show_bug.cgi?id=249207
rdar://103285803

Reviewed by Alan Baradlay.

Another patch will be added to provide support for block containers that
contain only inline level boxes.

There are 3 different pieces that we need to implement:

1.) Trimming the block start, inline start, and inline end margins
2.) Trimming the block-end margin, which is slightly different from
the others
3.) Making sure the trimmed margins do not contribute to the inrinsic
sizing of the container

When a candidate position for a float is determined in the float
positioning code, we can determine whether it is possible to trim its
margins. For each candidate position, we can determine whether any edges
of the margin box would touch the container and then trim that edge if
it is specified in margin-trim. This can be done for the inline-start,
inline-end, and block-start edges of the container.

Trimming the block-end margin is slightly different because it is not
in the context of the containing block like the other margins but of the
block formatting context. This means that we can only trim these margins
after its block formatting context is done with layout. In
RnderBlockFlow we can check to see if it establishes a block formatting
context and then proceed to trim the block-end margins of its floats
where needed. As we iterate over each float, we check to see if its block
end location is lower than the lowest piece of content in the BFC. If
so, we trim it up to the necesssary amount where it will not extend the
height of the BFC.

Finally, to make sure that the trimmed inline margins do not contribute
to the intrinsic sizing of its containing block, all we need to do
is check to see if the item is a float and if any of the inline margins
are specified to be trimmed. If so, we do not include those in its
contribution.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-deeply-nested-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-deeply-nested.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers-vert-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-different-containers.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-same-container-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-block-layout-same-container.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-relative-positioned-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-relative-positioned.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-with-transforms-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-with-transforms.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-fit-content-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-fit-content-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout-vert-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-left-trimmed-margin-allows-float-to-fit-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-with-block-content-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-with-block-content-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-right-trimmed-margin-allows-float-to-fit-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-right-trimmed-margin-allows-float-to-fit-block-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-overflowing-float-margins-tirmmed-at-final-position-block-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-overflowing-float-margins-tirmmed-at-final-position-block-layout.html: Added.
* Source/WebCore/rendering/FloatingObjects.cpp:
(WebCore::FloatingObject::isLowestPlacedFloatBottomInBlockFormattingContext const):
* Source/WebCore/rendering/FloatingObjects.h:
(WebCore::FloatingObject::setHeight):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeBlockPreferredLogicalWidths const):
* Source/WebCore/rendering/RenderBlock.h:
(WebCore::RenderBlock::logicalMarginBoxHeightForChild const):
(WebCore::RenderBlock::logicalMarginBoxTopForChild const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock):
(WebCore::RenderBlockFlow::trimFloatBlockEndMargins):
(WebCore::RenderBlockFlow::shouldChildInlineMarginContributeToContainerIntrinsicSize const):
(WebCore::RenderBlockFlow::insertFloatingObject):
(WebCore::RenderBlockFlow::trimMarginForFloat):
(WebCore::RenderBlockFlow::computeLogicalLocationForFloat):
(WebCore::RenderBlockFlow::blockFormattingContextInFlowBlockLevelContentHeight const):
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::establishesBlockFormattingContext const):
(WebCore::RenderBox::blockFormattingContextRoot const):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: https://commits.webkit.org/260318@main
